### PR TITLE
WIP, DONT MERGE

### PR DESCRIPTION
**Pull request by trancexpress**
Fixes: #1081

## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by trancexpress** [comment]
@jukzi one thing that remains the same is the `LookupEnvironment`.

I'm not sure this is the right fix, but lets see what the tests have to say.

I'm also not sure how we want to test this, as commented on #1081 so far I can reproduce the stack overflow only with an unbound classpath container. Will have to check how to specify an unbound container during tests...

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/LookupEnvironment.java [hunk 1]
```diff
@@ -333,12 +336,18 @@ ReferenceBinding askForType(PackageBinding packageBinding, char[] name, ModuleBi
...
+				boolean added = this.acceptingSourceFiles.add(fileName);
+				if (added) {
+					this.typeRequestor.accept(answer.getCompilationUnit(), answer.getAccessRestriction());
+				}
 			} catch (AbortCompilation abort) {
 				if (CharOperation.equals(name, TypeConstants.PACKAGE_INFO_NAME))
 					return null; // silently, requestor may not be able to handle compilation units (HierarchyResolver)
 				throw abort;
+			} finally {
+				this.acceptingSourceFiles.remove(fileName);
```
Comment (jukzi) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/LookupEnvironment.java:350]: need to remove only if added?
